src/thor/costmatrix.cc

```diff
@@ -262,21 +257,28 @@ void CostMatrix::SourceToTarget(Api& request,
     }
   }
 
+  // resize/reserve all properties of Matrix on first pass only
+  valhalla::Matrix& matrix = *request.mutable_matrix();
+  reserve_pbf_arrays(matrix, best_connection_.size(), costing_->pass());
+
   // Form the matrix PBF output
   graph_tile_ptr tile;
-  uint32_t count = 0;
-  valhalla::Matrix& matrix = *request.mutable_matrix();
-  reserve_pbf_arrays(matrix, best_connection_.size());
-  for (auto& connection : best_connection_) {
-    uint32_t target_idx = count % target_location_list.size();
-    uint32_t source_idx = count / target_location_list.size();
+  bool connection_failed = false;
+  for (uint32_t connection_idx = 0; connection_idx < best_connection_.size(); connection_idx++) {
+    auto best_connection = best_connection_[connection_idx];
+    // if this is the second pass we don't have to process previously found ones again
+    if (costing_->pass() > 0 && !(matrix.second_pass(connection_idx))) {
+      continue;
+    }
```

Comment:
if we're on the second pass and we didn't set the flag on the first pass, then don't bother processing anymore